syntax = "proto2";
package openscenario;

// Message for TrafficSignalAction
message TrafficSignalAction {
    optional TrafficSignalControllerAction trafficSignalControllerAction = 1;  // 0..1
    optional TrafficSignalStateAction trafficSignalStateAction = 2;  // 0..1
}

// Message for TrafficSignalControllerAction
message TrafficSignalControllerAction {
    required string phase = 1; 
    required TrafficSignalController trafficSignalControllerRef = 2;
    required Phase phaseRef = 3;
}

// Message for TrafficSignalStateAction
message TrafficSignalStateAction {
    required string name = 1;   // 1..1
    required string state = 2;  // 1..1
}

// Message for TrafficSignalGroupState
message TrafficSignalGroupState {
    required string state = 1;   // 1..1
}

// Message for TrafficSignalState
message TrafficSignalState {
    required string state = 1;   // 1..1
    required string trafficSignalId = 2;  // 1..1
}

// Message for TrafficSignalControllerCondition
message TrafficSignalControllerCondition {
    required string phase = 1;  // 1..1
    required TrafficSignalController trafficSignalControllerRef = 2;  // 1..1
    repeated Phase phaseRef = 3;  // 0..*
}

message TrafficSignals {
    repeated TrafficSignalController trafficSignalControllers = 1; // 0..*
}

// Message for TrafficSignalController
message TrafficSignalController {
    optional double delay = 1;   // 0..1
    required string name = 2;    // 1..1
    optional string reference = 3;  // 0..1
    repeated Phase phases = 4;   // 0..*
}

// Message for TrafficSignalCondition
message TrafficSignalCondition {
    required string name = 1;    // 1..1
    required string state = 2;   // 1..1
}

// Message for Phase
message Phase {
    required double duration = 1;   // 1..1
    required string name = 2;       // 1..1
    repeated TrafficSignalState trafficSignalStates = 3;  // 0..*
    optional TrafficSignalGroupState trafficeSignalGroupState = 4;  // 0..1
}
